### PR TITLE
[Mailer] Use AsyncAws to handle SES requests

### PR DESCRIPTION
**Pull request by jderusse**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | #33183, #35468 and #35037
| License       | MIT
| Doc PR        | TODO

alternative to #33326

This PR replace the native code to call AWS SES by the new [AsyncAws](https://github.com/async-aws/aws) project maintained by @Nyholm and me.

This removes complexity of signing request, and adds new features likes:
- authentication via .aws/config.ini, Instance profile, WebIdentity (K8S service account)
- usesignature V4 (the one recommanded by the Official SDK )
- fully compatible with API (uses the official AWS SDK interface contract to generate classes)

Because it's based on `symfony/http-client`, it's fully integrable with Symfony application.

### CODE REVIEW

File: src/Symfony/Component/Mailer/Bridge/Amazon/Tests/Transport/SesApiTransportTest.php [hunk 1]
```diff
@@ -34,19 +36,19 @@ public function getTransportData()
     {
         return [
             [
-                new SesApiTransport('ACCESS_KEY', 'SECRET_KEY'),
-                'ses+api://ACCESS_KEY@email.eu-west-1.amazonaws.com',
+                new SesApiTransport(new SesClient(Configuration::create(['accessKeyId' => 'ACCESS_KEY', 'accessKeySecret' => 'SECRET_KEY']))),
+                'ses+api://ACCESS_KEY@us-east-1',
```
Comment (jderusse) [inline src/Symfony/Component/Mailer/Bridge/Amazon/Tests/Transport/SesApiTransportTest.php]: Is it a BC Break?

File: src/Symfony/Component/Mailer/Bridge/Amazon/Tests/Transport/SesApiTransportTest.php [hunk 2]
```diff
@@ -56,28 +58,23 @@ public function testSend()
...
-</SendEmailResponse>';
+            $json = '{"MessageId": "foobar"}';
 
-            return new MockResponse($xml, [
+            return new MockResponse($json, [
                 'http_code' => 200,
             ]);
         });
-        $transport = new SesApiTransport('ACCESS_KEY', 'SECRET_KEY', null, $client);
-        $transport->setPort(8984);
```
Comment (jderusse) [inline src/Symfony/Component/Mailer/Bridge/Amazon/Tests/Transport/SesApiTransportTest.php]: changing the port without the host does not make sens

File: src/Symfony/Component/Mailer/Bridge/Amazon/Transport/SesApiTransport.php [hunk 3]
```diff
@@ -11,109 +11,95 @@
...
+        } catch (\Exception $e) {
+            throw new RuntimeException(sprintf('Unable to send message with the "%s" transport: %s', __CLASS__, $e->getMessage()), 0, $e);
+        }
 
-    private function getPayload(Email $email, Envelope $envelope): array
-    {
         if ($email->getAttachments()) {
-            return [
-                'Action' => 'SendRawEmail',
-                'RawMessage.Data' => base64_encode($email->toString()),
```
Comment (jderusse) [inline src/Symfony/Component/Mailer/Bridge/Amazon/Transport/SesApiTransport.php]: this is probably the issue of #35037

File: src/Symfony/Component/Mailer/Bridge/Amazon/composer.json [hunk 4]
```diff
@@ -20,6 +20,7 @@
         "symfony/mailer": "^4.4|^5.0"
     },
     "require-dev": {
+        "async-aws/ses": "^0.3",
```
Comment (jderusse) [inline src/Symfony/Component/Mailer/Bridge/Amazon/composer.json]: this is a BC break (people have to install this optional dependency) but it shouldn't be required IMO (one could use the bridge with SMTP=

File: src/Symfony/Component/Mailer/Bridge/Amazon/Transport/SesApiTransport.php [hunk 5]
```diff
@@ -11,109 +11,95 @@
...
         if ($emails = $email->getBcc()) {
-            $payload['Destination.BccAddresses.member'] = $this->stringifyAddresses($emails);
+            $destination->setBccAddresses($this->stringifyAddresses($emails));
         }
         if ($email->getTextBody()) {
-            $payload['Message.Body.Text.Data'] = $email->getTextBody();
+            $simpleMessage->setBody(new Body([
+                'Text' => [
+                    'Data' => $email->getTextBody(),
+                    'Charset' => $email->getTextCharset(),
```
Comment (jderusse) [inline src/Symfony/Component/Mailer/Bridge/Amazon/Transport/SesApiTransport.php]: this were not provided in current version

**Comment by chalasr** [comment]
IMHO AsyncAWS should also be a soft dependency when using the HTTP transport.
I think we would better preserve the current implementation and wire the AsyncAWS one when it is installed, since it is not a core dependency. 
Installing it by default could then be handled in a flex recipe.

**Comment by Nyholm** [comment]
Thank you for the PR and thank you @chalasr for the review. I agree with you. AsyncAws should be a soft dependency. Currently it is only installed in `require-dev` section. If you are trying to use features that require the external library you get an exception. See https://github.com/symfony/symfony/pull/35992/files#diff-4e7b1bdc9093065fae5fff42cdde5d8cR42

**Comment by chalasr** [comment]
@Nyholm My point is: we have an HTTP implementation that works out of the box currently. Requiring AsyncAWS to be available for using the HTTP transport implicitly makes it an hard dependency, even if it's not in the `require` section, since you now get an exception instead of something that works immediately.

Considering the pros + the fact that AsyncAWS is maintained by two core members, I'm OK for requiring it with a meaningful exception. 
But still, throwing in 5.1 is a BC break. IMHO we should deprecate the current implementation first, suggesting to install the missing package as it will be required in 6.0, and automatically wire the AsyncAWS-based implementation when it is installed.

**Comment by Nyholm** [comment]
Thank you. I understand the point you are trying to make. 

> we have an HTTP implementation that works out of the box currently.

Yes, But it has some bugs and it does not support all authentication mechanism available for AWS. 
Of course, bugs can be fixed and authentication can be implemented. 
We created AsyncAws to be able to abstract authentication away from this bridge. (And possible other parts of the symfony organization that may use AWS). We are also happy to extend the maintainer team with more Symfony core members if there is a wish for that. 

> But still, throwing in 5.1 is a BC break.

True, but this component and this bridge is still marked as experimental. 

-----------

Side note: Most code in Async AWS is generated so the maintenance is kept to a minimum.

**Comment by Nyholm** [comment]
> True, but this component and this bridge is still marked as experimental.

Oh no.. @chalasr pinged my privately and very politely pointed out that I was wrong. This component is not experimental.

**Comment by jderusse** [comment]
Thanks you for your review @chalasr. I agreed with you, The current HttpClient implementation should stay for BC.

But as @Nyholm pointed it, this implementation have several opened bugs. One of them is the way the request is signed. It currently use the version 2 which is not supported by all regions (https://docs.aws.amazon.com/general/latest/gr/signature-version-2.html). And maintaing a new implementation, is not trivial (see [async-aws](https://github.com/async-aws/aws/blob/master/src/Core/src/Signer/SignerV4.php) or [AWS SDK](https://github.com/aws/aws-sdk-php/blob/master/src/Signature/SignatureV4.php))

I suggest to use the AsyncAws Adapter when installed package is installed otherwise to fallback to HttpClient Adapter (as suggested by @chalasr), AND to deprecate the HttpClient Adapter in 5.1 to remove the complexity from Symfony.

WDYT?

**Comment by nicolas-grekas** [comment]
Since this is experimental, we can break BC and cut maintenance efforts.

**Comment by chalasr** [comment]
@nicolas-grekas Actually, this is not experimental :) https://github.com/symfony/symfony/pull/35992#issuecomment-599092314
@jderusse Sounds like a plan!

**Comment by jderusse** [comment]
PR updated according our discussion.

File: src/Symfony/Component/Mailer/Bridge/Amazon/CHANGELOG.md [hunk 6]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+5.1.0
+-----
+
+ * Adds `async-aws/ses` to communicate with AWS API.
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Amazon/CHANGELOG.md]: Add

File: composer.json [hunk 7]
```diff
@@ -126,7 +126,8 @@
         "phpdocumentor/reflection-docblock": "^3.0|^4.0",
         "twig/cssinliner-extra": "^2.12",
         "twig/inky-extra": "^2.12",
-        "twig/markdown-extra": "^2.12"
+        "twig/markdown-extra": "^2.12",
+        "async-aws/ses": "^0.3"
```
Comment (fabpot) [inline composer.json]: Will you have 1.0 ready before 5.1?

**Review by fabpot — CHANGES_REQUESTED** [review]

File: composer.json (hunk 7 above)
Comment (Nyholm) [reply composer.json]: Yes. That is the plan.

File: src/Symfony/Component/Mailer/Bridge/Amazon/CHANGELOG.md (hunk 6 above)
Comment (chalasr) [reply src/Symfony/Component/Mailer/Bridge/Amazon/CHANGELOG.md]: `Added` I'd say :)

Comment (jderusse) [reply src/Symfony/Component/Mailer/Bridge/Amazon/CHANGELOG.md]: fixed, thanks

**Comment by chalasr** [comment]
Tests seem broken

File: UPGRADE-6.0.md [hunk 8]
```diff
@@ -36,6 +36,12 @@ HttpFoundation
    `RedirectResponse::create()`, and `StreamedResponse::create()` methods (use
    `__construct()` instead)
 
+Mailer
+------
+
+ * Removed the `SesApiTransport` class. Run `composer require async-aws/ses` to keep the transport in your application.
```
Comment (stof) [inline UPGRADE-6.0.md]: this won't keep the removed transport class.

File: src/Symfony/Component/Mailer/Bridge/Amazon/Transport/SesTransportFactory.php [hunk 9]
```diff
@@ -11,35 +11,65 @@
...
-            return (new SesHttpTransport($user, $password, $region, $this->client, $this->dispatcher, $this->logger))->setHost($host)->setPort($port);
-        }
+        if (!class_exists(SesClient::class)) {
+            if (!class_exists(HttpClient::class)) {
+                throw new \LogicException(sprintf('You cannot use "%s" as the HttpClient component or AsyncAws package is not installed. Try running "composer require async-aws/ses".', __CLASS__));
+            }
 
-        if ('ses+smtp' === $scheme || 'ses+smtps' === $scheme) {
-            return new SesSmtpTransport($user, $password, $region, $this->dispatcher, $this->logger);
+            trigger_deprecation('symfony/amazon-mailer', '5.1', 'Using the "%s" transport without AsyncAws is deprecated, use "new %s()" instead.', __METHOD__, \get_called_class());
```
Comment (stof) [inline src/Symfony/Component/Mailer/Bridge/Amazon/Transport/SesTransportFactory.php]: `use new SesTransportFactory()` looks wrong to me in the message.

Comment (jderusse) [reply src/Symfony/Component/Mailer/Bridge/Amazon/Transport/SesTransportFactory.php]: thanks, reworded

**Comment by chalasr** [comment]
Rebase needed

**Comment by connorhu** [comment]
@jderusse Since this is an alternative solution, can I close the #33326 pull request?

**Comment by Nyholm** [comment]
@connorhu Probably, but let's wait to see if this PR is accepted. If this PR is rejected then #33326 might be something that should be considered.

**Review by Nyholm — APPROVED** [review]
If this is ready for merge we should release 1.0.0 of async-aws/ses. 

@jderusse Could you verify the failing tests?

**Review by chalasr — APPROVED** [review]
CI failures are unrelated. 👍

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @jderusse.

**Pull request merged**